src/App.tsx

```diff
@@ -1,24 +1,295 @@
+        setErrorMessage('Unable to add a todo');
+      })
+      .finally(() => {
+        setIsLoadingAddTodo(false);
+        setTempTodo(null);
+      });
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      });
+  }, []);
+
+  const handelDeleteTodoById = (todoId: number) => {
+    setTodosIdToDelete(prevState => [...prevState, todoId]);
+
+    deleteTodoById(todoId)
+      .then(() => {
+        setTodos(prevState => prevState.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => {
+        setTodosIdToDelete(prevState => prevState.filter(id => id !== todoId));
+      });
+  };
+
+  const handleChangeCompletedStatus
+    = (todoId: number, isCompleted: boolean) => {
+      setTodosIdToUpdate(prevState => [...prevState, todoId]);
+
+      updateTodoById(todoId, { completed: !isCompleted })
+        .then(() => {
+          setTodos(prevState => {
+            const currentTodo = prevState
+              .find(todo => todo.id === todoId);
+
+            if (!currentTodo) {
+              return prevState;
+            }
+
+            const newTodos = [...prevState];
+            const index = prevState.indexOf(currentTodo);
+
+            newTodos.splice(index, 1, {
+              ...currentTodo,
+              completed: !currentTodo.completed,
+            });
+
+            return newTodos;
+          });
+        })
+        .catch(() => {
+          setErrorMessage('Unable to update a todo');
+        })
+        .finally(() => {
+          setTodosIdToUpdate(prevState => prevState
+            .filter(id => id !== todoId));
+        });
+    };
+
+  const handleChangeTodoTitle
+    = (todoId: number, newTitle: string) => {
+      setTodosIdToUpdate(prevState => [...prevState, todoId]);
+
+      updateTodoById(todoId, { title: newTitle })
+        .then(() => {
+          setTodos(prevState => {
+            const currentTodo = prevState
+              .find(todo => todo.id === todoId);
+
+            if (!currentTodo) {
+              return prevState;
+            }
+
+            const newTodos = [...prevState];
+            const index = prevState.indexOf(currentTodo);
+
+            newTodos.splice(index, 1, {
+              ...currentTodo,
+              title: newTitle,
+            });
+
+            return newTodos;
+          });
+        })
+        .catch(() => {
+          setErrorMessage('Unable to update a todo');
+        })
+        .finally(() => {
+          setTodosIdToUpdate(prevState => prevState
+            .filter(id => id !== todoId));
+        });
+    };
+
+  const handelDeleteCompletedTodos = async () => {
+    const completedTodosId = todos
+      .filter(({ completed }) => completed)
+      .map(({ id }) => id);
+
+    await Promise
+      .all([...completedTodosId.map(id => handelDeleteTodoById(id))]);
```

Comment:
Great 🔥 
But don't forget to wrap it into `try/catch` block

Reply:
why would I do try/catch if I already have error handling in handelDeleteTodoById?